### PR TITLE
feat: add Context Caching to Gemini plugins

### PR DESCRIPTION
**Pull request by cabljac**
Checklist (if applicable):
- [x] Tested - manually
- [ ] Changelog updated
- [x] Docs updated

### CODE REVIEW

File: js/plugins/googleai/src/gemini.ts [hunk 1]
```diff
@@ -631,15 +638,54 @@ export function defineGoogleAIModel(
...
         safetySettings: requestConfig.safetySettings,
       } as StartChatParams;
-      const msg = toGeminiMessage(messages[messages.length - 1], model);
 
-      const fromJSONModeScopedGeminiCandidate = (
-        candidate: GeminiCandidate
-      ) => {
-        return fromGeminiCandidate(candidate, jsonMode);
-      };
+      let cache: CachedContent | null = null;
```
Comment (cabljac) [inline js/plugins/googleai/src/gemini.ts:642]: can we avoid this

File: js/plugins/googleai/src/gemini.ts [hunk 2]
```diff
@@ -631,15 +638,54 @@ export function defineGoogleAIModel(
...
+
+      // TODO: fix casting
```
Comment (cabljac) [inline js/plugins/googleai/src/gemini.ts:644]: fix todo

File: js/plugins/googleai/src/gemini.ts [hunk 3]
```diff
@@ -631,15 +638,54 @@ export function defineGoogleAIModel(
...
+          request,
+          chatRequest,
+          modelVersion,
+          cacheConfigDetails
+        );
+        chatRequest = handleContextCacheResponse.newChatRequest;
+        cache = handleContextCacheResponse.cache;
+      }
+
+      let genModel: GenerativeModel | null = null;
```
Comment (cabljac) [inline js/plugins/googleai/src/gemini.ts:666]: fix this

File: js/plugins/googleai/src/context-caching/utils.ts [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+};
+
+/**
+ * Validates context caching request for compatibility with model and request configurations.
+ */
+export function validateContextCacheRequest(
+  request: any,
+  modelVersion: string
+): boolean {
+  if (!modelVersion || !CONTEXT_CACHE_SUPPORTED_MODELS.includes(modelVersion)) {
```
Comment (cabljac) [inline js/plugins/googleai/src/context-caching/utils.ts:195]: this probably shouldn't throw - we should probably just return false

Comment (cabljac) [reply js/plugins/googleai/src/context-caching/utils.ts:195]: It should probably throw if they *try* to use cache (cache is in metadata etc) but the model version is wrong. We don't want to accidentally cost them lots of tokens.

File: js/plugins/googleai/src/context-caching/utils.ts [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    pageToken = nextPageToken;
+    currentPage++;
+  }
+  return null;
+}
+
+/**
+ * Clears all caches using the cache manager.
+ */
+export async function clearAllCaches(
```
Comment (cabljac) [inline js/plugins/googleai/src/context-caching/utils.ts:121]: this isnt actually used

File: docs/plugins/google-genai.md [hunk 6]
```diff
@@ -133,3 +133,158 @@ const llmResponse = await ai.generate({
...
+    {
+      role: 'model',
+      content: [
+        {
+          text: 'Based on War and Peace, here is some analysis of Pierre Bezukhov’s character.',
+        },
+      ],
+      metadata: {
+        cache: {
+          ttlSeconds: 300, // Cache this message for 5 minutes
```
Comment (pavelgj) [inline docs/plugins/google-genai.md:169]:
```suggestion
          ttlSeconds: 300, // Cache all the message up to and including this message for 5 minutes
```

File: docs/plugins/vertex-ai.md [hunk 7]
```diff
@@ -421,4 +421,94 @@ See the code samples for:
...
+    {
+      role: 'model',
+      content: [
+        {
+          text: 'Based on War and Peace, here is some analysis of Pierre Bezukhov’s character.',
+        },
+      ],
+      metadata: {
+        cache: {
+          ttlSeconds: 300, // Cache this message for 5 minutes
```
Comment (pavelgj) [inline docs/plugins/vertex-ai.md:456]:
```suggestion
          ttlSeconds: 300, // Cache all messages up to and including this message for 5 minutes
```

File: docs/plugins/vertex-ai.md [hunk 8]
```diff
@@ -421,4 +421,94 @@ See the code samples for:
...
+});
+```
+
+In this setup:
+- **`messages`**: Allows you to pass conversation history.
+- **`metadata.cache.ttlSeconds`**: Specifies the time-to-live (TTL) for caching a specific response.
+
+### Example: Leveraging Large Texts with Context
+
+For applications referencing long documents, such as *War and Peace* or *Lord of the Rings*, you can structure your queries to reuse cached contexts:
```
Comment (pavelgj) [inline docs/plugins/vertex-ai.md:472]:
```suggestion
For applications referencing long documents, ex. the size of *War and Peace* or *Lord of the Rings*, you can structure your queries to reuse cached contexts:
```

File: js/plugins/googleai/src/context-caching/utils.ts (hunk 5 above)
Comment (pavelgj) [reply js/plugins/googleai/src/context-caching/utils.ts:121]: remove?

**Review by pavelgj — APPROVED** [review]

**Pull request merged**